rfcs/20240311-gather-scatter-batching-dims.md

```diff
@@ -0,0 +1,427 @@
+* `update_window_index = update_index[update_window_dims...]`.
+* `full_window_index = [wi0, ..., 0, ..., wiN]` where `wi` are individual
+  elements in `update_window_index`, and `0` is inserted at indices from
+  `inserted_window_dims` and `input_batching_dims`.
+* `result_index = full_start_index + full_batching_index + full_window_index`.
+
+Given that, `results = exec(schedule, inputs)`, where:
+
+* `schedule` is an implementation-defined permutation of
+  `index_space(updates[0])`.
+* `exec([update_index, ...], results) = exec([...], updated_results)` where:
+  * If `result_index` is in bounds for `shape(results...)`
+    * `updates_converted = to_destination_type(
+      updates...[update_index], type(func_inputs(update_computation)
+      [len(func_inputs(update_computation))//2:])... )`
+    * `updated_values = update_computation(results...[result_index], updates_converted)`
+    * `updated_results` is a copy of `results` with `results...[result_index]`
+      set to `updated_values...`.
+  * Otherwise
+    * `updated_results = results`.
+* `exec([], results) = results`.
+
+If `indices_are_sorted` is `true` then the implementation can assume that
+`scatter_indices` are sorted with respect to `scatter_dims_to_operand_dims`,
+otherwise the behavior is undefined. More formally, for all `i1 < i2` from
+`indices(result)`, `full_start_index(i1)` <= `full_start_index(i2)`.
+
+If `unique_indices` is `true` then the implementation can assume that all
+`result_index` indices being scattered to are unique. If `unique_indices` is
+`true` but the indices being scattered to are not unique then the behavior is
+undefined.
+
+#### Inputs
+
+| Label | Name                                  | Type                                                       | Constraints                                         |
+|-------|---------------------------------------|------------------------------------------------------------|-----------------------------------------------------|
+| (I1)  | `inputs`                              | variadic number of tensors or per-tensor quantized tensors | (C1), (C2), (C4-C6), (C11), (C13), (C21), (C23-C24) |
+| (I2)  | `scatter_indices`                     | tensor of integer type                                     | (C4), (C15), (C19), (C22)                           |
+| (I3)  | `updates`                             | variadic number of tensors or per-tensor quantized tensors | (C3-C6), (C8)                                       |
+| (I4)  | `update_window_dims`                  | 1-dimensional tensor constant of type `si64`               | (C2), (C4), (C7-C8)                                 |
+| (I5)  | `inserted_window_dims`                | 1-dimensional tensor constant of type `si64`               | (C2), (C4), (C9-C11)                                |
+| (I6)  | `input_batching_dimensions`           | 1-dimensional tensor constant of type `si64`               | (C2), (C4), (C9), (C12-13), (C17-18), (C20)         |
```

Comment:
Do we want to name this as `inputs_batching_dimensions` (note the extra `s` after `input`) for consistency?

Reply:
I'm not sure, because the dimensions are of each input, which is why I named it input (singular). Is this for consistency with `inputs`? I'm happy to change if you think this is preferable.

Reply:
Keeping it as-is is fine.

Originally, I thought we should simply add `_batching_dimensions` to `inputs` and `scatter_indices`.  But the plurality of `inputs` refers to multiple input tensors, where the plurality of `scatter_indices`  refers to the indices within a tensor.  So using singular makes sense too.

Reply:
SG :)